qrexec/tools/qrexec_legacy_convert.py

```diff
@@ -0,0 +1,309 @@
+
+    # all rules that exist only in legacy files
+    legacy_rules = [rule for rule in all_rules if rule not in new_rules]
+
+    # all services for which a legacy rule exists
+    all_services = {rule.service for rule in legacy_rules}
+
+    # dict of file_name: rules list
+    rules_to_save = {CONFIG_FILE: []}
+    for filename in SERVICE_TO_FILE.values():
+        rules_to_save[filename] = []
+
+    for service in all_services:
+        legacy = [rule for rule in legacy_rules if rule.service == service]
+        non_legacy = [rule for rule in new_rules if rule.service == service]
+
+        legacy_str = [str(rule) for rule in legacy]
+        non_legacy_str = [str(rule) for rule in non_legacy]
+
+        if legacy_str == non_legacy_str or not legacy:
+            continue
+
+        missing = [rule for rule in legacy if str(rule) not in non_legacy_str]
+        if not missing:
+            continue
+
+        last_working_rule = len(missing)
+        for i, rule in enumerate(missing):
+            if str(rule.rule.action) == 'deny' \
+                    and str(rule.rule.source) == '@anyvm' \
+                    and str(rule.rule.target) == '@anyvm' \
+                    and not rule.rule.argument:
+                last_working_rule = i
+                break
+        missing = missing[:last_working_rule]
+        if not missing:
+            continue
+        filename = SERVICE_TO_FILE.get(service, CONFIG_FILE)
+
+        for rule in missing:
+            if rule.is_rule_compatible():
+                rules_to_save[filename].append(rule)
+            else:
+                rules_to_save[CONFIG_FILE].append(rule)
+
+    if '50-config-input' in rules_to_save:
+        config_file, user_file = \
+            split_input_rules(rules_to_save['50-config-input'])
+        rules_to_save['50-config-input'] = config_file
+        rules_to_save[CONFIG_FILE].extend(user_file)
+
+    backups_made: Dict[pathlib.Path, pathlib.Path] = {}
+
+    # do actual rule saving
+    for filename, rules in rules_to_save.items():
+        if not rules:
+            continue
+
+        file = POLICYPATH / (filename + ".policy")
+        disclaimer = DISCLAIMER if filename == '30-user' else TOOL_DISCLAIMER
+        text = disclaimer + '\n'.join([str(rule.rule) for rule in rules]) + '\n'
+
+        if file.exists():
+            backup_name = str(file) + '.bak'
+            while pathlib.Path(backup_name).exists():
+                backup_name = backup_name + '.bak'
+            backups_made[file] = pathlib.Path(backup_name)
+            shutil.copy(file, backup_name)
+
+            if filename != '50-config-input':
+                # input files are special: they replace current rules
+                current_text = file.read_text()
+                if current_text.startswith(TOOL_DISCLAIMER):
+                    current_text = current_text[len(TOOL_DISCLAIMER):]
+                text = text.rstrip('\n') + current_text
```

Comment:
The rstrip here is wrong at least if the old file doesn't start with an empty line. It will cause last line of converted rules be concatenated with first line of pre-existing rules -> syntax error.